interface/ceed-basis.c

```diff
@@ -636,6 +636,44 @@ int CeedHouseholderApplyQ(CeedScalar *mat_A, const CeedScalar *mat_Q, const Ceed
   return CEED_ERROR_SUCCESS;
 }
 
+/**
+  @brief Return pseudoinverse of a matrix
+
+  @param[in]     ceed Ceed context for error handling
+  @param[in]     mat  Row-major matrix to be factorized in place
+  @param[in]     m    Number of rows
+  @param[in]     n    Number of columns
+  @param[out]    mat_pinv  Row-major pseudoinverse matrix
+
+  @return An error code: 0 - success, otherwise - failure
+
+  @ref Utility
+**/
+int CeedMatrixPseudoinverse(Ceed ceed, CeedScalar *mat, CeedInt m, CeedInt n, CeedScalar *mat_pinv) {
+  CeedScalar *tau, *I;
+
+  CeedCall(CeedCalloc(m * n, &I));
+  CeedCall(CeedCalloc(m, &tau));
+  // -- QR Factorization, mat = Q R
+  CeedCall(CeedQRFactorization(ceed, mat, tau, m, n));
+  // -- mat_pinv = R_inv Q^T
+  for (CeedInt i = 0; i < m; i++)
+    for (CeedInt j = 0; j < n; j++) I[i * m + j] = 1.0;
+  // ---- Apply R_inv, mat_pinv = I R_inv
+  for (CeedInt i = 0; i < m; i++) {  // Row i
+    mat_pinv[m * i] = I[m * i] / mat[0];
+    for (CeedInt j = 1; j < n; j++) {  // Column j
+      mat_pinv[j + m * i] = I[j + m * i];
+      for (CeedInt k = 0; k < j; k++) mat_pinv[j + m * i] -= mat[j + m * k] * mat_pinv[k + m * i];
+      mat_pinv[j + m * i] /= mat[j + n * j];
+    }
+  }
+  // ---- Apply Q^T, mat_pinv = R_inv Q^T
+  CeedCall(CeedHouseholderApplyQ(mat_pinv, mat, tau, CEED_NOTRANSPOSE, m, n, n, 1, m));
+
+  return CEED_ERROR_SUCCESS;
+}
```

Comment:
Am I using the indices in a correct way here? I tried to use this function in `CeedBasisGetCollocatedGrad` as
```c
CeedCall(CeedMatrixPseudoinverse(ceed, interp_1d, Q_1d, P_1d, interp_1d_pinv));
  CeedCall(CeedMatrixMatrixMultiply(ceed, (const CeedScalar *)interp_1d_pinv, (const CeedScalar *)grad_1d, collo_grad_1d, Q_1d, P_1d, P_1d));
```
but breaks some tests.

Reply:
I think you want to swap the order of the MatMatMult - grad_1d * interp_1d_pinv = collograd_1d

Reply:
Yeah, I tried that, didn't work. In last commit I changed the indices and I tested t302-basis.c , still doesn't work but I don't get memory error at least.

Reply:
I remembered why I don't like this code, lol. I did some fiddling and this might work?

```diff
diff --git a/interface/ceed-basis.c b/interface/ceed-basis.c
index 60c7a6e9..8e5d304a 100644
--- a/interface/ceed-basis.c
+++ b/interface/ceed-basis.c
@@ -259,26 +259,21 @@ static int CeedBasisCreateProjectionMatrices(CeedBasis basis_from, CeedBasis bas
 **/
 int CeedBasisGetCollocatedGrad(CeedBasis basis, CeedScalar *collo_grad_1d) {
   Ceed        ceed;
-  CeedInt     P_1d = (basis)->P_1d, Q_1d = (basis)->Q_1d;
-  CeedScalar *interp_1d, *interp_1d_pinv, *grad_1d, *tau;
+  CeedInt     P_1d, Q_1d;
+  CeedScalar *interp_1d_pinv;
 
-  CeedCall(CeedMalloc(Q_1d * P_1d, &interp_1d));
-  CeedCall(CeedMalloc(P_1d * Q_1d, &interp_1d_pinv));
-  CeedCall(CeedMalloc(Q_1d * P_1d, &grad_1d));
-  CeedCall(CeedMalloc(Q_1d, &tau));
-  memcpy(interp_1d, (basis)->interp_1d, Q_1d * P_1d * sizeof(basis)->interp_1d[0]);
-  memcpy(grad_1d, (basis)->grad_1d, Q_1d * P_1d * sizeof(basis)->interp_1d[0]);
-
-  // QR Factorization, interp_1d = Q R
   CeedCall(CeedBasisGetCeed(basis, &ceed));
+  CeedCall(CeedBasisGetNumNodes1D(basis, &P_1d));
+  CeedCall(CeedBasisGetNumQuadraturePoints1D(basis, &Q_1d));
 
-  CeedCall(CeedMatrixPseudoinverse(ceed, interp_1d, Q_1d, P_1d, interp_1d_pinv));
-  CeedCall(CeedMatrixMatrixMultiply(ceed, (const CeedScalar *)interp_1d_pinv, (const CeedScalar *)grad_1d, collo_grad_1d, Q_1d, P_1d, P_1d));
+  // QR Factorization, interp_1d = Q R
+  CeedCall(CeedCalloc(P_1d * Q_1d, &interp_1d_pinv));
+  CeedCall(CeedMatrixPseudoinverse(ceed, basis->interp_1d, Q_1d, P_1d, interp_1d_pinv));
+  CeedCall(CeedMatrixMatrixMultiply(ceed, basis->grad_1d, (const CeedScalar *)interp_1d_pinv, collo_grad_1d, Q_1d, Q_1d, P_1d));
 
-  CeedCall(CeedFree(&interp_1d));
+  // Cleanup
   CeedCall(CeedFree(&interp_1d_pinv));
-  CeedCall(CeedFree(&grad_1d));
-  CeedCall(CeedFree(&tau));
+
   return CEED_ERROR_SUCCESS;
 }
 
@@ -627,36 +622,45 @@ int CeedHouseholderApplyQ(CeedScalar *mat_A, const CeedScalar *mat_Q, const Ceed
 /**
   @brief Return pseudoinverse of a matrix
 
-  @param[in]     ceed Ceed context for error handling
-  @param[in]     mat  Row-major matrix to be factorized in place
-  @param[in]     m    Number of rows
-  @param[in]     n    Number of columns
-  @param[out]    mat_pinv  Row-major pseudoinverse matrix
+  @param[in]  ceed     Ceed context for error handling
+  @param[in]  mat      Row-major matrix to be factorized in place
+  @param[in]  m        Number of rows
+  @param[in]  n        Number of columns
+  @param[out] mat_pinv Row-major pseudoinverse matrix
 
   @return An error code: 0 - success, otherwise - failure
 
   @ref Utility
 **/
 int CeedMatrixPseudoinverse(Ceed ceed, CeedScalar *mat, CeedInt m, CeedInt n, CeedScalar *mat_pinv) {
-  CeedScalar *tau, *I;
+  CeedScalar *tau, *I, *mat_copy;
 
-  CeedCall(CeedCalloc(m * m, &I));
   CeedCall(CeedCalloc(m, &tau));
-  // -- QR Factorization, mat = Q R
-  CeedCall(CeedQRFactorization(ceed, mat, tau, m, n));
-  // -- mat_pinv = R_inv Q^T
-  for (CeedInt i = 0; i < m; i++) I[i * m + i] = 1.0;
-  // ---- Apply R_inv, mat_pinv = I R_inv
-  for (CeedInt i = 0; i < m; i++) {  // Row i
-    mat_pinv[m * i] = I[n * i] / mat[0];
-    for (CeedInt j = 1; j < n; j++) {  // Column j
-      mat_pinv[j + m * i] = I[j + n * i];
-      for (CeedInt k = 0; k < j; k++) mat_pinv[j + m * i] -= mat[j + n * k] * mat_pinv[k + m * i];
-      mat_pinv[j + m * i] /= mat[j + n * j];
+  CeedCall(CeedCalloc(n * n, &I));
+  CeedCall(CeedCalloc(m * n, &mat_copy));
+  memcpy(mat_copy, mat, m * n * sizeof mat[0]);
+
+  // QR Factorization, mat = Q R
+  CeedCall(CeedQRFactorization(ceed, mat_copy, tau, m, n));
+
+  // mat_pinv = R_inv Q^T
+  for (CeedInt i = 0; i < n; i++) I[i * n + i] = 1.0;
+  // -- Apply R_inv, mat_pinv = I R_inv
+  for (CeedInt i = 0; i < n; i++) {  // Row i
+    mat_pinv[n * i] = I[n * i] / mat_copy[0];
+    for (CeedInt j = 1; j < m; j++) {  // Column j
+      mat_pinv[j + n * i] = I[j + n * i];
+      for (CeedInt k = 0; k < j; k++) mat_pinv[j + n * i] -= mat_copy[j + m * k] * mat_pinv[k + n * i];
+      mat_pinv[j + n * i] /= mat_copy[j + n * j];
     }
   }
-  // ---- Apply Q^T, mat_pinv = R_inv Q^T
-  CeedCall(CeedHouseholderApplyQ(mat, mat_pinv, tau, CEED_NOTRANSPOSE, m, n, n, 1, n));
+  // -- Apply Q^T, mat_pinv = R_inv Q^T
+  CeedCall(CeedHouseholderApplyQ(mat_pinv, mat_copy, tau, CEED_NOTRANSPOSE, n, m, n, 1, n));
+
+  // Cleanup
+  CeedCall(CeedFree(&I));
+  CeedCall(CeedFree(&tau));
+  CeedCall(CeedFree(&mat_copy));
 
   return CEED_ERROR_SUCCESS;
 }
```

I ended up pulling out pencil and paper and checking matrix dimensions.

If this works, then next fix would be for the projection function.

Reply:
Thanks, yeah it works.

Reply:
In projection we ApplyQ with transpose mode in the beginning
```c
  for (CeedInt m = 0; m < num_matrices; m++) {
    // Apply Q^T, interp_from = Q^T interp_from
    memcpy(interp_from, input_from[m], Q * P_from * q_comp * sizeof(input_from[m][0]));
    CeedCall(CeedHouseholderApplyQ(interp_from, interp_to, tau, CEED_TRANSPOSE, Q * q_comp, P_from, P_to, P_from, 1));

    // Apply Rinv, output_project = Rinv interp_from
    for (CeedInt j = 0; j < P_from; j++) {  // Column j
      output_project[m][j + P_from * (P_to - 1)] = interp_from[j + P_from * (P_to - 1)] / interp_to[P_to * P_to - 1];
      for (CeedInt i = P_to - 2; i >= 0; i--) {  // Row i
        output_project[m][j + P_from * i] = interp_from[j + P_from * i];
        for (CeedInt k = i + 1; k < P_to; k++) {
          output_project[m][j + P_from * i] -= interp_to[k + P_to * i] * output_project[m][j + P_from * k];
        }
        output_project[m][j + P_from * i] /= interp_to[i + P_to * i];
      }
    }
  }
```

It is like a reverse of what we wrote in `CeedMatrixPseudoinverse` since we ApplyQ with no transpose mode at the end.

Reply:
ASAN is saying that this fix is still wrong

Reply:
Yeah, that patch is wrong and makes NaNs

Reply:
Yeah, I just checked `t302-basis.c` which was fine I need to write it down on paper.

Reply:
Alternatively, we could just swap to your previous suggestion
```
/**
  @brief Compute A B_pinv = C

  @param[in]  ceed   Ceed context for error handling
  @param[in]  A      First matrix
  @param[in]  B      Matrix to apply pesudoinverse of
  @param[in]  t_mode A B_pinv (NOTRANSPOSE) or B_pinv A (TRANSPOSE)
  @param[in]  m      Number of rows in C
  @param[in]  n      Number of columns in C
  @param[in]  k      Number of colums in A and B (NOTRANSPOSE)
  @param[out] C      Row-major pseudoinverse matrix

  @return An error code: 0 - success, otherwise - failure

  @ref Utility
**/
int CeedMatrixMatrixPseudoinverseMultiply(Ceed ceed, CeedScalar *A, CeedScalar *B, CeedTransposeMode t_mode, CeedInt m, CeedInt n, CeedInt k, CeedScalar *C) {
```
or similar?

Reply:
I don't see an obvious reason why this isn't working but it seems to be some stupid indexing issue.

Reply:
Yeah there is an issue with indices. At the beginning I considered `interp_1d` with size `QxP` and its inverse `interp_1d_pinv` size is `PxQ` so I said if we want to create `grad_1d` with size `QxP` we need `interp_1d x interp_1d_pinv`

Reply:
The collocated gradient has size QxQ

Reply:
Ah, yeah. I'll work on it.